src/Processors/Formats/Impl/IonRowOutputFormat.cpp

```diff
@@ -0,0 +1,323 @@
+            size_t offset = offsets[row_num - 1];
+            size_t size = offsets[row_num] - offset;
+            writer->writeStartStruct();
+            for (size_t i = 0; i < size; ++i)
+            {
+                if (map_type.getKeyType()->getTypeId() == TypeIndex::String)
+                    writer->writeStructFieldName(assert_cast<const ColumnString &>(*key_column).getDataAt(offset + i).toView());
+                else if (map_type.getKeyType()->getTypeId() == TypeIndex::FixedString)
+                    writer->writeStructFieldName(assert_cast<const ColumnFixedString &>(*key_column).getDataAt(offset + i).toView());
+                else
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Ion only supports string types for map keys");
+                serializeField(*value_column, map_type.getValueType(), offset + i);
+            }
+            writer->writeFinishStruct();
+            return;
+        }
+        case TypeIndex::LowCardinality: {
+            const auto & lc_column = assert_cast<const ColumnLowCardinality &>(column);
+            auto dict_type = assert_cast<const DataTypeLowCardinality *>(data_type.get())->getDictionaryType();
+            auto dict_column = lc_column.getDictionary().getNestedColumn();
+            size_t index = lc_column.getIndexAt(row_num);
+            serializeField(*dict_column, dict_type, index);
+            return;
+        }
+        case TypeIndex::UUID: {
+            const auto & uuid_column = assert_cast<const ColumnUUID &>(column);
+            WriteBufferFromOwnString buf;
+            writeText(uuid_column.getElement(row_num), buf);
+            writer->writeString(buf.stringView());
+            return;
+        }
+        default:
+            break;
+    }
+    throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Type {} is not supported for MsgPack output format", data_type->getName());
+}
+
+void IonRowOutputFormat::writeRowStartDelimiter()
+{
+    writer->writeStartStruct();
+}
+
+void IonRowOutputFormat::writeRowEndDelimiter()
+{
+    writer->writeFinishStruct();
+}
+
+void IonRowOutputFormat::writePrefix()
+{
+    writer->writeStartList();
+}
+
+void IonRowOutputFormat::writeSuffix()
+{
+    writer->writeFinishList();
+}
+
+void IonRowOutputFormat::finalizeImpl()
+{
+    writer->final();
+}
+
+void IonRowOutputFormat::write(const Columns & columns, size_t row_num)
+{
+    writeRowStartDelimiter();
+    for (size_t i = 0; i < columns.size(); ++i)
+    {
+        writer->writeStructFieldName(names[i]);
+        serializeField(*columns[i], types[i], row_num);
+    }
+    writeRowEndDelimiter();
+}
+
+void registerOutputFormatIon(FormatFactory & factory)
+{
+    factory.registerOutputFormat(
+        "Ion",
+        [](WriteBuffer & buf, const Block & sample, const FormatSettings & settings)
+        { return std::make_shared<IonRowOutputFormat>(buf, sample, settings); });
+    factory.markOutputFormatSupportsParallelFormatting("Ion");
```

Comment:
Did you check that parallel formatting actually works? Need to add a test. See example in `tests/queries/1_stateful/00159_parallel_formatting_csv_and_friends.sh`